[AMDGPU] Emit a waitcnt instruction after each memory instruction

jwanggit86 opened the pull request:
This patch implements a new command-line option for the backend, namely, amdgpu-waitcnt-for-all-mem-op. When this option is specified, a "waitcnt 0" instruction is generated after each memory load/store instruction.

jwanggit86 committed:
[AMDGPU] Emit a waitcnt instruction after each memory instruction

This patch implements a new command-line option for the backend, namely,
amdgpu-waitcnt-for-all-mem-op. When this option is specified, a "waitcnt 0"
instruction is generated after each memory load/store instruction.

jayfoad commented:
What's the reason for this?

Do we really need both this option and the existing `-amdgpu-waitcnt-forcezero`?

jwanggit86 commented:
A feature request asks for this because it would make debugging memory problems easier.

jayfoad commented:
> A feature request asks for this because it would make debugging memory problems easier.

OK, I guess I can see that it might be useful to put the s_waitcnt immediately after the thing that "generates" the wait. `-amdgpu-waitcnt-forcezero` is different because it still puts the s_waitcnt before the thing that "consumes" the wait, but it forces the waitcnt value to be 0 instead of something higher.

I would prefer it if your implementation was a bit more integrated with the rest of the pass, instead of just adding a separate pass over all instructions. You might be able to do this by adding a call to `generateWaitcnt` near the end of `insertWaitcntInBlock`, after the call to `updateEventWaitcntAfter`. One advantage of doing it this way is that it would work for all wait counters including things like EXP_CNT which apply to non-memory instructions.

jwanggit86 commented:
Thanks for the comments. Doing the insertion of "waitcnt 0" in `insertWaitcntInBlock`, however, would mean that we have to check the newly-created flag (`EmitForAllMemOpFlag`) for _each instruction_. The proposed code change, on the other hand, checks this flag only once for each function. Because this new command-line option is for debug purposes only and not used in normal cases, we probably should limit the extra overhead for normal uses.

arsenm commented:
I think it would be better to refine the existing flag to control the behavior. It could expand to an enum for wait before/wait after.

Also on some hardware, gdb can force this behavior anyway

jwanggit86 commented:
The waitcnt insertions by the SIInsertWaitcnt pass is not controlled by any command-line option. In other words, there's no command-line option to enable/disable the insertions. There is one option, however, i.e., `--amdgpu-waitcnt-forcezero`, when used, would force all the inserted waitcnt to be "waitcnt 0". On the other hand, the command-line option for this feature is an enable/disable type of option. So, while it's possible to create a combined option, using a separate option also makes sense. Do we generally try to avoid creating new command-line options?

jwanggit86 committed:
(1) change cmdline option name to amdgpu-precise-memory-op (2) option
is added for both clang and llc

jwanggit86 committed:
Remove some obsolete code.

jwanggit86 committed:
Fix code formatting for GCNSubtarget.h.

jwanggit86 committed:
Fix code format.

jwanggit86 committed:
Fix formatting.

arsenm commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -52,6 +52,11 @@ static cl::opt<bool> ForceEmitZeroFlag(
   cl::desc("Force all waitcnt instrs to be emitted as s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)"),
   cl::init(false), cl::Hidden);
 
+static cl::opt<bool>
+    PreciseMemOpFlag("amdgpu-precise-memory-op",
+                     cl::desc("Emit s_waitcnt 0 after each memory operation"),
+                     cl::init(false));
+
```
I think this should be fused into an enum flag with the existing waitcnt flag.

jwanggit86 replied:
Sure. I will look into this.

jwanggit86 replied:
Adding a command-line option for llc is from the 1st version of this patch. Now that we have decided to add a clang command-line option, this llc option is not really necessary anymore.

arsenm commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -1809,6 +1816,23 @@ bool SIInsertWaitcnts::shouldFlushVmCnt(MachineLoop *ML,
+bool SIInsertWaitcnts::insertWaitcntAfterMemOp(MachineFunction &MF) {
+  bool Modified = false;
+
+  for (auto &MBB : MF) {
```
Should try to integrate with the rest of the logic instead of adding a separate pass over the function

jwanggit86 replied:
Although they both insert s_waitcnt instructions, the new feature is quite different from the existing SIInsertWaitcnt pass. The new feature, controlled by a command-line option, inserts a "s_waitcnt 0" after each memory instruction. The logic therefore is very simple. The existing pass, however, has more complicated logic implemented with essentially a static analysis aided by its own data structures, which are not necessary for the new feature.

From the performance point of view, it should be noted that by default this feature is not activated. Therefore, extra overhead should be minimized for the normal use-case scenario. A separate pass achieves this b/c there is only one extra IF for each compiled function. On the other hand, integrating with the existing pass would mean many more checks for the feature activation, which are waste in the normal case when the feature is not activated.

With the above 2 points, I think a separate pass is advantageous over an integrated pass. Pls let me know your thoughts.

arsenm replied:
I think it makes it harder to reason about the pass as a whole to have it as a totally separate phase

jwanggit86 replied:
As commented above, the newly created llc option is not really necessary and will be removed.

jwanggit86 replied:
@arsenm Pull req has been updated. Now the code for the new feature has been integrated with the existing pass.

jwanggit86 committed:
Remove the command-line option for llc because it is not really
necessary,given that a command-line option has been added for clang.

jwanggit86 committed:
Integrating inserting "s_waitcnt 0" with existing SIInsertWaintcnt
pass.

arsenm commented:
> So, while it's possible to create a combined option, using a separate option also makes sense. Do we generally try to avoid creating new command-line options?

Looking again, I see they are different and unrelated. I don't really understand why we have amdgpu-waitcnt-forcezero, I'm not sure I've ever used it. I always expected it to behave like this flag.

jwanggit86 commented:
> > So, while it's possible to create a combined option, using a separate option also makes sense. Do we generally try to avoid creating new command-line options?
> 
> Looking again, I see they are different and unrelated. I don't really understand why we have amdgpu-waitcnt-forcezero, I'm not sure I've ever used it. I always expected it to behave like this flag.

So do you still think the new code should be integrated with the existing pass, or should it be separate?

jwanggit86 commented:
@arsenm Pls let me know if you have any further comments. I'd like to have the code committed and the PR closed. Thanks!

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -388,6 +388,8 @@ class SIInsertWaitcnts : public MachineFunctionPass {
   // message.
   DenseSet<MachineInstr *> ReleaseVGPRInsts;
 
+  // bool insertWaitcntAfterMemOp(MachineFunction &MF);
```
Remove all the unused code, don't just comment it out.

jwanggit86 replied:
Done.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -1708,6 +1710,13 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
 
     ++Iter;
+    if (ST->isPreciseMemoryEnabled() && Inst.mayLoadOrStore()) {
+      auto builder =
```
Upper case B for Builder.

jwanggit86 replied:
Done.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -1708,6 +1710,13 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
+    if (ST->isPreciseMemoryEnabled() && Inst.mayLoadOrStore()) {
+      auto builder =
+          BuildMI(Block, Iter, DebugLoc(), TII->get(AMDGPU::S_WAITCNT))
+              .addImm(0);
```
On GFX10+ VMEM stores should have S_WAITCNT_VSCNT 0 as well as (or instead of) this.

jwanggit86 replied:
Done.

jayfoad commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_all.ll:
```diff
@@ -0,0 +1,222 @@
+; Testing the -amdgpu-precise-memory-op option
+; COM: llc -mtriple=amdgcn -mcpu=hawaii -mattr=+amdgpu-precise-memory-op -verify-machineinstrs < %s | FileCheck %s -check-prefixes=GFX7
```
What is COM: ?

jwanggit86 replied:
Comment. Some testcases in this file won't run if mcpu=hawaii. In the latest commit, the test file has been split into 2.

jwanggit86 committed:
Made changes based on code review: (1) for GFX10+ generate a
s_waitcnt_vscnt 0 as well (2) split test files into two. (3) removed
unused code.

jwanggit86 committed:
Fix code formatting.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -1708,6 +1710,19 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
+                .addReg(AMDGPU::SGPR_NULL, RegState::Undef)
+                .addImm(0);
+      }
+      OldWaitcntInstr = Builder.getInstr();
```
Nit: if you're going to set OldWaitcntInstr then really it ought to point to the first in a sequence of waitcnts, not the last.

jwanggit86 replied:
Done.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -1708,6 +1710,19 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
+      auto Builder =
+          BuildMI(Block, Iter, DebugLoc(), TII->get(AMDGPU::S_WAITCNT))
+              .addImm(0);
+      if (IsGFX10Plus) {
```
I guess this works but it seems a bit wasteful to insert S_WAITCNT after stores and S_WAITCNT_VSCNT after loads. Does anyone care?

Stepping back a bit, I think you can probably implement this by calling generateWaitcnt instead of building the instructions yourself.

jwanggit86 replied:
S_waitcnt(0) is inserted after each mem op, both stores and loads.

jayfoad replied:
Yes but why? On GFX10+, why would you put s_waitcnt(0) after a store or s_waitcnt_vscnt(0) after a load?

jwanggit86 replied:
My understanding is that the feature request asks for a "s_waitcnt 0" to be *blindly* inserted after each and every memory instruction. Enabling the feature is at the user's discretion via a clang command-line option (disabled by default). The purpose of the feature is to help debug memory problems on GPUs that do not support precise memory. (Although someone, Tony I think, mentioned it could go beyond debugging). I'll send you the link for the feature request.

Based on that, the implementation doesn't check on GPU models, doesn't have model-dependent code (except the newly-added code for GFX10+), or differentiate loads from stores. I'll work with the requester to get the requirements straightened out.

jwanggit86 committed:
Code changed based on code review: (1) update OldWaitcntInstr such that
it points to the inserted s_waitcnt even when the s_waitcnt is followed
by an s_waitcnt_vscnt (2) updated testcases accordingly.

jwanggit86 commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -1847,6 +1862,7 @@ bool SIInsertWaitcnts::runOnMachineFunction(MachineFunction &MF) {
 
   TrackedWaitcntSet.clear();
   BlockInfos.clear();
+
```
Done.

jwanggit86 commented:
Implementation is moved to SIMemoryLegalizer pass. See pull req [79236](https://github.com/llvm/llvm-project/pull/79236).
